Computer Science: HashMap Project (#26518)
TheOdinProject/curriculum: javascript/computer_science/project_hash_map.md

### Introduction

One of the most used data structure across all languages is a Hash Table, aka Hash Map. If you worked before with JavaScript Object Literal `{}`, `Set`, and `Map` then you are already familiar with a hash table. But how do they work exactly, how can we save values using strings and symbols as keys and then retrieving those values using the same keys.

> In this project, you will learn how it all works, and you will implement your own Hash Map! To start we need to learn a few concepts, one of which is a hash code, what does it mean, and how to use it. Let us get started and hope you are exited to finally know what is happening with an object when you put keys and values in them.

Reviewer: There are a bunch of different ways "hash map" is written in this lesson. It appears as "hash map", "Hash Map", "HashMap", etc.

I think it'd be good to have a consistent approach. My preference would probably be "hash map" and then if we're referring to the class the user is making in the project, maybe `HashMap` can be fine too.

### What is a hash code

Let us start by learning what does it mean to hash a value. Hashing means taking an input in and generating an output that represents this input. A Hash function should be a pure function, hashing the same input should always return the same hash code, there should be no random generation that is happening. For example, let's work a hashing function that takes a name and gives us the first letter of that name:

```javascript
function hash(name) {
  return name.charAt(0);
}
```

> We created our first basic hashing function, it takes a name and returns the first letter of that name.

Reviewer: ```suggestion
We created our first basic hashing function. It takes a name and returns the first letter of that name.
```

> You might be familiar with this concept from your cipher exercise earlier in foundation, but there is a key difference between hashing and ciphering (encryption), which is reversibility.

Reviewer: ```suggestion
You might be familiar with this concept from your cipher exercise earlier in the foundations course, but there is a key difference between hashing and ciphering (encryption): reversibility.
```

> Hashing is a one-way process. Here is an example given a name, you can get a hash out of it, but given a hash, you cannot revert it back to a name. If you have a name `"Carlos"` we can hash it to `"C"`. But it will not be possible to reverse it from `"C"` back to its original form. You cannot know if it is `"Carlos"`, maybe it's `"Carla"` or `"Carrot"` we don't know. This is very good for security also, given a password you can save the hash of that password, but if someone steals your hashes they cannot know the original passwords since they are unable to reverse the hash back to the password. We are not going to dive deeper into hashing passwords, this is a subject for another lesson.

Reviewer: ```suggestion
Hashing is a one-way process. In the above example, you can get a hash code from a name, but when given a hash, you cannot revert it back to a name. If you have a name `"Carlos"` we can hash it to `"C"`. But it will not be possible to reverse it from `"C"` back to its original form. You cannot know if the original name was `"Carlos"`. Maybe it's `"Carla"` or `"Carrot"`. We can't know. This is very good for security. Given a password you can save the hash of that password, but if someone steals your hashes they cannot know the original passwords because they are unable to reverse the hash back to the password. We are not going to dive deeper into hashing passwords, and this is a subject for another lesson.
```
I also think the information about passwords might fit better as a note.

### Use cases

> What can we do with those hashes, You have probably seen it in school where a folder is organized into smaller folders, each folder holds information about people with the same first letter, so we get something like this:

Reviewer: ```suggestion
What can we do with those hashes? You have probably seen it in school where a folder is organized into smaller folders, and each folder holds information about people with the same first letter, so we get something like this:
```

```
C:
  carlos.txt
  carla.txt
B:
  bryan.txt
  bob.txt
  beatrice.txt
  bella.txt
  benjamin.txt
  bianca.txt
```

> If we get new student in our school with the name `"Carlos"` we can run our hash function to find out what folder do we assign them in `hash("Carlos") -> "C"` so we put `"Carlos"` in the directory labeled `C`

Reviewer: ```suggestion
If we get new student in our school with the name `"Carlos"`, we can run our hash function to find out which folder to place them in. `hash("Carlos") -> "C"` so we put `"Carlos"` in the directory labeled `C`
```

You might have spotted a problem, what if our school is populated with many people that share the same first letter `C`? Then we will have a directory with labeled `C` that have too many names that start with `"C"` and the other directories are empty. We need to rework our hash function, to eliminate more duplication and separate our students, while keeping constant structure in folders.

```javascript
function hash(name, surname) {
  return name.charAt(0) + surname.charAt(0);
}
```

> Instead of just taking the first name letter, we take the first name and last name letters. `"Carlos Smith"` will have a hash code of `"CS"` this will surely spread our students in more directories, it will eliminate many duplicate hash codes from being generated.

Reviewer: ```suggestion
Instead of just taking the first name letter, we take the first name and last name letters. `"Carlos Smith"` will have a hash code of `"CS"`. This will spread our students among more directories and will eliminate many duplicate hash codes from being generated.
```

![Example of hashing a name using first name's first letter, and last name's first letter](./project_hash_map/imgs/00.png)

But it still doesn't solve our problem, what if we have a very popular first name's first letter and surname's first letter? Then we are again with directories that are empty and other directories that are full, we need it easier for us to find the person we are looking for, so let's rework our hash code.

```javascript
function stringToNumber(string) {
  let hashCode = 0;
  for (let i = 0; i < string.length; i++) {
    hashCode += string.charCodeAt(i);
  }

  return hashCode;
}

function hash(name, surname) {
  return stringToNumber(name) + stringToNumber(surname);
}
```

> Using this technique we don't only take the first letters into consideration, now we take the whole name and convert it into numbers.

Reviewer: ```suggestion
We not only consider the first letters with this technique. Instead, we take the entire name and convert it into numbers.
```

> You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true, this would make it unique for each name, but there are other reasons that we have chosen a number:

Reviewer: ```suggestion
You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true. This would make it unique for each name, but there are some good reasons to choose a number:
```

This is it, making this will result in a hash table with `search`, `set` and `get`.

What if we found the hash code, but also the key value is the same as what we already have in the bucket. We check if it's the same item by comparing the keys, then we overwrite the value with our new value. This is how we can only have unique values inside a `Set`, `Set` is similar to a hash map but the key difference (pun intended) is that a `Set` will have Nodes with only keys and no values.

<div class="lesson-note lesson-note--warning" markdown="1">
A Hash Map does not guarantee insertion order, when you iterate over it. That is because of the buckets indexing nature of hash coding. That means if you are to retrieve the array of keys and values to iterate over them, then they will not be in order of when you inserted them.
For example, if we insert the values `Mao`, `Zach`, `Xari` in this order, we may get back `["Zach", "Mao", "Xari"]`
If iterating over the hash map frequently is your goal, then this data structure is not the right choice for the job, a simple array would be better.

> </div>

Author: I thought about adding reserving order as a bonus assignment, but not sure if that will deter people from the main concept of hash maps.

- Insertion
- Retrieving
- Deletion

Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.

The Growth of our hash map has the complexity of `O(n)` at all times.

### Assignment

<div class="lesson-content__panel" markdown="1">

#### Self restriction and limitation

  Before we get started, we need to lay down some ground rules. As we all know JavaScript is a dynamic language the restriction we have in other languages does not exist in JavaScript for example if you create an array of size `16` to be your buckets size, nothing stopping you from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.

  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.

      ```javascript
      if (index < 0 || index >= buckets.length) {
        throw new Error("Trying to access index out of bound");
      }
      ```

>   1. Inside the class DO NOT use Objects to store your key value pairs, And DO NOT use the array method `push` to insert your nodes into the buckets, always use indexes.

Author: > Also, is it okay to store the key/value pair as an object on the linked list nodes?

Since Node is not part of the hash map and using an object literal is not violating the rule, but I still feel that this takes the magic away. In the end of the day, using this will still make us give our key values to JS built in library to handle for us. That's what we're trying to avoid, and building our own key value mechanics.

#### Time to create a real hash map

>   1. Start by creating a HashMap class or factory function, up to you. Then proceed to create the following methods:

Reviewer: ```suggestion
  1. Start by creating a `HashMap` class or factory function. It's up to you which you want to use. Then proceed to create the following methods:
```

>   1. `hash` takes a value and produces a hash code with it. There are a lot of thought that goes into creating the most optimal hashing function, it is up to you if you want to do a research to find a good hashing function but do not dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes that will lead to too many collisions.

Reviewer: ```suggestion
  1. `hash` takes a value and produces a hash code with it. There is a lot of thought that goes into creating optimal hashing functions. It is up to you if you want to research to find a good hashing function, but don't dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes. This will lead to frequent collisions when storing keys.
```

Author: Can you go over this again, please? I don't feel this present the idea that if we have a function that produces similar keys, then it will lead to frequent collisions. It seems you separated the concerns. Or we can re-arrange it to talk about frequent collisions earlier and that we need to avoid them.

Reviewer: I like this change, and think it's better than mine :laughing: @01zulfi 

Do you have thoughts on it @Mclilzee 

Another thing I think we may want to specify here is: what types of things we can hash and use as keys in the hash map. With ruby hashes and JavaScript `Map`, you can use just about anything as a key (numbers, arrays, sets, objects). Under the hood, the hashing algorithm will generate different codes for things like `1` vs. `"1".

The other way would be like a JavaScript object implementation, where everything is just coerced to strings.
```js
const obj = {}
obj[1] = "value"; // stored a number key right

console.log(obj["1"]); // wrong
// logs "value"
```
Since the lesson only talked about how to deal with hashing strings, I think taking the 2nd approach would be best? And I also think it's probably a good idea to say that people will not be required to deal with hashing different types.

Author: It is still the same concept, the only thing that changes is the hash code input, the output is still a number. at this point, students should have enough experience to work around this.

We can add it as extra bonus to update the hash function so that it accepts any object and be able to generate meaningful hash code with it, Thoughts?

Reviewer 2: IMO this could be a limitation as the project does seem quite lengthy as it stands. Something like:

> Hash maps could accommodate various data types for keys like numbers, strings, objects. But for this project, only handle keys of type strings.

  1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key.

>      - Remember to grow your buckets size when it needs to, by calculating the `load factor` and `capacity`.

Reviewer: I think another good reminder here is to think about what needs to happen when the given key already exists in the hash map.

This can be a bit of a balance. I'm sure some things we want them to "discover" while doing the project, but with this, I think it's entirely plausible that a learner does the project without testing or thinking about what happens when you `set()` with a key that's already there. So I think it'd be good to call attention to this case.

Author: I put it as part of the lesson that if a key already present, then its value is overwritten. I can add another small reminder here, like the one I put for growing the table.

>   1. `get` takes one argument as a key and returns the value that is assigned to this key.

Reviewer: What should happen when the given key doesn't exist in the hash? When a function has a return value, we should be explicit about the different possible returns. Just like with the [Linked List assignment](https://www.theodinproject.com/lessons/javascript-linked-lists#assignment) for example

>   1. `has` takes a key as an argument and checks if your hash map has that key.

Reviewer: ```suggestion
  1. `has` takes a key as an argument and returns `true` or `false` based on whether or not the key is in the hash map.
```

>   1. `remove` takes a key as argument and remove it from the hash table.

Reviewer: ```suggestion
  1. `remove` takes a key as argument and removes it from the hash table.
```

>   1. `remove` takes a key as argument and remove it from the hash table.

Reviewer: Should `remove()` return anything? Usually it does return something, if for no other reason than to let the developer know if an item has been deleted or not.

[JavaScript Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value) returns `true` if the key was deleted and `false` if it wasn't.

In Ruby hashes, [the value for the deleted item is returned](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value) (or `nil` if the key wasn't found).

Either one of those approaches could make sense to me. Maybe with this JS project, we can design it like the JS `Map`?

Author: This is true, I don't think I have come across a language that returns nothing with its remove function. Good catch, I think returning true or false will suffice for the JS implementation, maybe it can be reworked in Ruby lesson to be more similar to the hash map in ruby where it returns `nil` if not found? That part is up to you

>   1. `length` returns how many stored keys in the hash code.

Reviewer: ```suggestion
  1. `length` returns the number of stored keys in the hash map.
```

  1. `clear` removes all entries in the hash map.

  1. `keys` returns an array containing all the keys inside the hash map.

  1. `values` returns an array containing all the values.

>   1. `entries` returns an array that contains each a `key, value` pairs. Example: `[[firstKey, firstValue], [secondKey, secondValue]]`

Reviewer: ```suggestion
  1. `entries` returns an array that contains each `key, value` pair. Example: `[[firstKey, firstValue], [secondKey, secondValue]]`
```

In this project, you will learn how it all works, and you will implement your own hash map! To start we need to learn a few concepts, one of which is a hash code, what does it mean, and how to use it. Let us get started and hope you are exited to finally know what is happening with an object when you put keys and values in them.

### What is a hash code

Let's start by learning what it means to hash a value. Hashing involves taking an input in and generating an output that represents the initial input. A hash function should be a pure function. Hashing the same input should always return the same hash code, and there should be no random generation component. For example, let's look at a hashing function that takes a name and gives us the first letter of that name:

```javascript
function hash(name) {
  return name.charAt(0);
}
```

We created our first basic hashing function. It takes a name and returns the first letter of that name.

> You might be familiar with this concept from your cipher exercise earlier in the foundations course, but there is a key difference between hashing and ciphering (encryption): reversibility.

Author: The concept of ciphering, where you take an input and produce an output that represent that input, like how people did it in the Caesar cipher exercise. Which I have just realized that it has been archived…

<div class="lesson-note lesson-note--tip" markdown="1">
Hashing is very good for security, given a password you can save the hash of that password, but if someone steals your hashes they cannot know the original passwords since they are unable to reverse the hash back to the password.
</div>

### Use cases

What can we do with those hashes? You have probably seen it in school where a folder is organized into smaller folders, and each folder holds information about people with the same first letter, so we get something like this:

```
C:
  carlos.txt
  carla.txt
B:
  bryan.txt
  bob.txt
  beatrice.txt
  bella.txt
  benjamin.txt
  bianca.txt
```

If we get new student in our school with the name `"Carlos"`, we can run our hash function to find out which folder to place them in. `hash("Carlos") -> "C"` so we put `"Carlos"` in the directory labeled `C`

> You might have spotted a problem: what if our school is populated with many people whose names share the same first letter `C`? Then we will have a directory labeled `C` that holds too many names that start with `"C"` while the other directories are empty. To eliminate this duplication and better separate our students, we need to rework our hash function while still maintaining a constant folder structure.

Author: Without changing the way we arrange our folders, we just update how we're going to label and insert items into them. But if it is not clear to you, then it probably not clear for others. I will remove that sentence

```javascript
function stringToNumber(string) {
  let hashCode = 0;
  for (let i = 0; i < string.length; i++) {
    hashCode += string.charCodeAt(i);
  }

  return hashCode;
}

function hash(name, surname) {
  return stringToNumber(name) + stringToNumber(surname);
}
```

We not only consider the first letters with this technique. Instead, we take the entire name and convert it into numbers.

You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true. This would make it unique for each name, but there are some good reasons to choose a number:

- We do not save the hash code. The hash code is a calculation to find out which bucket (storage) our value will be placed in. Think about it as a locker number for example.

- We save computation time. If we are to find where our hash is stored, it would be much faster and easier for a computer to compare two numbers rather than a string (or objects) to find the bucket where we have to store our elements (you will learn more about buckets shortly). Finding a bucket using a number would allow us to use that number as our index.

> - By using number as hash code, we can have hash codes that is uniform across variables where it doesn't have to only be a string, we can even hash numbers, or objects then we need an algorithm to generate a hash code number for our needs from an object, taking `class Person` object as an example, our hash code can be a combination of `strinngToNummber(name) + age`.

Author: At this point I wanted the ease in the switch I made after explaining hash codes using letters, and now we switched to a number, wanted to elaborate to not make that switch feel forced. But I may have written too many distracting words from the point of this lesson too, I'm not against this

### Buckets

Buckets are storage that we need to store our elements. Simply, it's an array. From now on, we will be referring to our storage as buckets. We have a set of buckets as our storage, since our hash function now produce a number we are going to find out which bucket we gonna use for storage using that number as index. We receive a key `"Fred"` We hash it using our hash function, it produces the number `508` we look which Bucket has index `508` we put `"Fred"` in the bucket. This is the simple form, but there are more mechanics that we need to deal with.

To get a value using a key, we put each entry inside a bucket as a `Node` item, which hold both the key and the value. To retrieve the value we hash a key, we find the bucket number, if the bucket is not empty then we go to that bucket we compare if the Node's key is the same key that is already in the bucket, if it is then we retrieve the Node's value otherwise we return null.

> Maybe you are wondering, why are we comparing the keys if we already found the hash code of that bucket? That is because remember, hash code is just the location, different items might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.

Author: You mean to replace the Bucket section entirely with a note? I believe it's too lengthy to be looked at as a note, unless we shorten it down. I don't think it hurts to keep it as a section, it fits nicely with all the bucket foreshadowing we have been doing in earlier sections.

> Maybe you are wondering, why are we comparing the keys if we already found the hash code of that bucket? That is because remember, hash code is just the location, different items might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.

Reviewer 2: > Maybe you are wondering, why are we comparing the keys if we already found the hash code of that bucket? That is because remember, hash code is just the location, different items might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.

^ I just mean this

> This is it, making this will result in a hash table with `search`, `set` and `get`.

Author: I believe I meant `has` not `search` my bad

What if we found the hash code, but also the key value is the same as what we already have in the bucket. We check if it's the same item by comparing the keys, then we overwrite the value with our new value. This is how we can only have unique values inside a `Set`, `Set` is similar to a hash map but the key difference (pun intended) is that a `Set` will have Nodes with only keys and no values.

<div class="lesson-note lesson-note--warning" markdown="1">
A hash map does not guarantee insertion order when you iterate over it, Because of the buckets indexing nature of hash coding. That means if you are to retrieve the array of keys and values to iterate over them, then they will not be in order of when you inserted them.

Some libraries implement hash tables with insertion order in mind such as JavaScript's own `Map`. For this project however we will be implementing an unordered hash table.
Example: if we insert the values `Mao`, `Zach`, `Xari` in this order, we may get back `["Zach", "Mao", "Xari"]` when we call an iterator.

If iterating over the hash map frequently is your goal, then this data structure is not the right choice for the job, a simple array would be better.

> </div>

Author: I wanted to explain that since hash codes translation to indexes does not go in linear fashion from first to last index, and rather more chaotic of where the index going to be no matter the insert order.

If we are to add a heading, wouldn't it be better to remove the note entirely and just make it a subsection?

> </div>

Reviewer 2: > hash codes translation to indexes does not go in linear fashion from first to last index, and rather more chaotic of where the index going to be no matter the insert order.

this is more clear than "indexing nature of hash coding", can you add this to the lesson instead please?

> If we are to add a heading, wouldn't it be better to remove the note entirely and just make it a subsection?

Agreed.

Turn out, we can rework our `stringToNumber` function so that it can give us unique hash codes which depends on where the letter appear in the name using an algorithm.

```javascript
function stringToNumber(string) {
  let hashCode = 0;

  const primeNumber = 31;
  for (let i = 0; i < string.length; i++) {
    hashCode = primeNumber * hashCode + string.charCodeAt(i);
  }

  return hashCode;
}
```

With our new function we will have different hash codes for the names `"Sara"` and `"raSa"` that is because even if both names have the same letters, some of the letters appear in different locations. The hash code started to change because we are multiplying the old hash every new iteration and then adding the letter code.

<div class="lesson-note lesson-note--tip" markdown="1">
  Notice the usage of prime number. We could have chosen any number we wanted, but prime numbers are even better to introduce less hash codes that are divisible by the same bucket length, which will make collisions less likely to happen.

> </div>

Author: It's how I learnt it the first time, long ago, from some sources and video on YouTube.

You might want to explore this lengthy discussion https://stackoverflow.com/questions/299304/why-does-javas-hashcode-in-string-use-31-as-a-multiplier/299748

> </div>

Reviewer 2: Cool read; this link should definitely be in the lesson

> Even though we reworked our hash function to avoid the `"Sara"/"raSa"` collision, there is always the possibility for collisions. Because we have a finite number of buckets, there is no way to eliminate collisions entirely. But we can try to minimize them as much as possible.

Author: I believe you missed a `we` after Since there, I will add this in my next PR

#### Dealing with collisions

> Enters `Linked Lists`. If each `Node` inside the bucket is also a Linked List, Then we look for bucket `508` if it's empty we insert the head of Linked List, If a Node head exists in a bucket we follow that Linked List to add to the end of it.

Reviewer 2: Wording suggestion:
```suggestion
Up until now, our hash map is a one-dimensional data structure. What if each `Node` inside the bucket can store more than one value? Enters `Linked Lists`. Now, each `Node` inside the bucket is also a Linked List. When inserting to a bucket, if it's empty, we insert the head of Linked List. If a `Node` head exists in a bucket, we follow that Linked List to add to the end of it.
```

You probably understand by this point why we must write a good hashing function which eliminates as many collisions as possible. Most likely you will not be writing your own hash functions, as most languages have it built in. But understanding how hash functions work is important.

> Check out [This Video](https://www.youtube.com/watch?v=btT4bCOvqjs) from CS50 that explains the concept shortly with visualization

Reviewer 2: This resource should go in the assignment section of the lesson.

### Growth of a hash table

Let's talk about the growth of our buckets. We don't have infinite memory, so we can't have an infinite amount of buckets. We need to start somewhere, but starting too big is also a waste of memory if we're only going to have a hash map with a single value in it. So to deal with this issue, we should start with a small array for our buckets. We'll use an array with 10 elements and indexes from 0 to 9.

<div class="lesson-note lesson-note--tip" markdown="1">
  Most programming languages start with the default size of `16` because it's a power of 2, which help with some techniques for performance that require bit manipulation for indexes. But for this example, we will be using a starting size of 10.

> </div>

Reviewer 2: Nice addition. I wonder if we can also just use 16 as the starting size of our array?

How are we going to insert into those buckets when our hash function generates big numbers like `20353924`? We make use of the modulo `%` operation `given any number modulo by 10 we will get a number between 0 and 9`.
For example, if we are to find the bucket where the value `"Manon"` will land, then we do the following:

> ![hashing using hash code and modular operation example](./project_hash_map/imgs/01.png)

Reviewer 2: This paragraph doesn't contribute to the "Growth of hash tables" section. Maybe move it to "Buckets" section?

Author: The Buckets section does not introduce the initial size of our array, I had to add this to clear a confusion of how are we going to translate big numbers when our array is only limited to 16? If I need to move this to the buckets section, then I will have to find a way to introduce the array size there as well.

Any suggestions?

Reviewer 2: Nice observation. Then, I propose the following wording suggestion:
```diff
- How are we going to insert into those buckets when our hash function generates big numbers like `20353924`? We make use of the modulo `%` operation `given any number modulo by 10 we will get a number between 0 and 9`.
+ Now that we have imposed the restriction of finite buckets, how will we deal with hash codes that exceed the available indexes in our hash map like `20353924`.  We make use of the modulo `%` operator. 
```

> If we keep adding nodes into our buckets then the buckets will start filling up, but what is more important is we know for a fact that if almost all buckets have items in them, then it is guaranteed that some buckets will have collisions, It is Mathematically impossible not to.

Reviewer 2: Wording suggestion:
```suggestion
If we keep adding nodes into our buckets then the buckets will start filling up, but what is more important is we know for a fact that if almost all buckets have items in them, then collisions are guaranteed, it is mathematically impossible not to.
```

> Remember we don't want collisions, in a perfect world each bucket will either have 0 or 1 Node only, so we grow our buckets to have more chance that our Nodes will spread and not stack up in the same buckets. To grow our buckets, we create a new buckets list that is double the size of the old buckets list, then we retrieve all nodes from the old buckets and insert them into the new buckets.

Reviewer 2: Wording suggestion: 
```suggestion
Remember we don't want collisions, in a perfect world each bucket will either have 0 or 1 Node only, so we grow our buckets to have more chance that our Nodes will spread and not stack up in the same buckets. To grow our buckets, we create a new buckets list that is double the size of the old buckets list, then we copy all nodes over to the new buckets.
```

> #### When do we know that it's time to grow our buckets size

Reviewer 2: Love this section and how in-depth the lesson goes on discussing the inner workings of hash map 💪

To deal with this, our hash map class need to keep track of two new fields, the `capacity` and the `load factor`.

> - The `capacity` is the amount of buckets we currently have. Keeping track of this will let us know if our map has reached the threshold, the threshold will depend on our factor variable as well.

Reviewer 2: I'm not sure what this means: "the threshold will depend on our factor variable as well." We could eliminate this phrase entirely, thoughts?
```suggestion
- The `capacity` is the total number of buckets we currently have. Keeping track of this will let us know if our map has reached a certain threshold aka `load factor`,
```

> - The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have collisions before we grow them. Usually a good balance of `0.75 to 1` is used.

Reviewer 2: Wording suggestion: 
```suggestion
- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have many collisions before we grow them. Usually a good balance of `0.75 to 1` is used.
```

### Computation complexity

A hash map is very efficient in its insertion, retrieving and removing operations. That is because we use array indexes to do those operations. A hash map has an average case complexity of `O(1)` for the following methods:
- Insertion
- Retrieving
- Deletion

Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.

> The Growth of our hash map has the complexity of `O(n)` at all times.

Reviewer 2: I feel this section doesn't talk about how collisions and use of linked lists affect complexities

Author: I don't follow. The complexity itself surfaces because of the linked list, and `O(n)` is because we are traversing the linked list to insert yet again another Node in the same bucket. Which happens specifically because of collisions.

Any suggestions?

Reviewer 2: > The complexity itself surfaces because of the linked list, and O(n) is because we are traversing the linked list to insert yet again another Node in the same bucket. Which happens specifically because of collisions.

^ this is exactly what I think should be included when talking about the worst case

### Introduction

> It is time that you create your own hash map. You already know the magic of how it all works out from the previous lesson, now it is time for you to create your own magical implementation.

Reviewer 2: Wording suggestion: 
```suggestion
You already know the magic behind hash maps, now it's time to write your own implementation!
```

#### Limitation

  Before we get started, we need to lay down some ground rules. JavaScript's dynamic nature of array allow us to insert and retrieve indexes that is outside our array size range. Example: if we create an array of size `16` to be our buckets size, nothing stopping us from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.

  1. Use the following snippet whenever you access a bucket through an index. We want to throw an error if we try to access an out of bound index:

      ```javascript
      if (index < 0 || index >= buckets.length) {
        throw new Error("Trying to access index out of bound");
      }
      ```

>   1. Inside the class DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.

Reviewer 2: Won't our learners be implementing linked lists for buckets? So I don't think we have to worry about them using the `push` method.

Author: Linked Lists are the items inside the buckets. The buckets themselves needs to be created using an Array and be accessed through indexes only.

Author: That leads me to believe your suggestion for this to be changed was because you believed Linked List were the buckets themselves? 
https://github.com/Mclilzee/curriculum/commit/490c30f43b7ca387c33b47eedc8ddd086c22bde8#diff-2ae1edcda41d250feb4bded479b31f662bd5c7231c17553e9e351b9ba9e02b67L137

I will revert it back so that it is more clear that Nodes as linked lists are _inside_ the buckets

Reviewer 2: I'm not married to any semantic here, just stating what is clear to _me_. If you feel saying "linked list is _inside_ a bucket" is clearer, go with that. I wonder if @JoshDevHub has a preference here.

Author: I assumed you thought about buckets being implemented in a Linked List and learners have no access to the `push` method. But since we are on the same page, I don't want learners to use `push` to insert the nodes inside the buckets.

I see your point of view more clearly now, where if you have a linked list inside the buckets, then the push method will not help since you need to traverse that list to find a location for a new node.

>   1. Inside the class DO NOT use Objects to store your key value pairs, and DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.

Reviewer: I think he means using object literals like:
```js
// not okay
{ key: keyToInsert, value: valueToInsert }

// :thumbsup:
new Node(keyToInsert, valueToInsert)  // class
Node(keyToInsert, valueToInsert)  // factory
```
I agree that we need to be precise with the language here. JavaScript mingles the idea of "hash map" and "object" in a way most other languages don't.

>   1. Inside the class DO NOT use Objects to store your key value pairs, and DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.

Author: It could be renamed to Object Literal `{}` and can be more specific on how not to use that object literal, Josh showcase of what is ok and what is not ok is brilliant we can use that

>   1. Inside the class DO NOT use Objects to store your key value pairs, and DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.

Reviewer 2: It's clearly said in the lesson that key value pairs are stored in the Linked List's node data structure. Then, why do we have to state that objects cannot be used when the learner will always be using a node data structure.

<div class="lesson-content__panel" markdown="1">

  1. Start by creating a `HashMap` class or factory function. It's up to you which you want to use. Then proceed to create the following methods:

  1. `hash` takes a value and produces a hash code with it. There are a lot of thought that goes into creating the most optimal hashing function, it is up to you if you want to do a research to find a good hashing function but do not dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes that will lead to too many collisions.

  1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key, if a key already exists then the old value is overwritten.

     - Remember to grow your buckets size when it needs to, by calculating if your bucket has reached the `load factor`.

  1. `get` takes one argument as a key and returns the value that is assigned to this key. If key is not found, return `null`.

  1. `has` takes a key as an argument and returns `true` or `false` based on whether or not the key is in the hash map.

  1. `remove` takes a key as argument and removes it from the hash table.

>   1. `length` returns the number of stored keys in the hash map.

Author: Capacity is inner details that don't usually get exposed to the outside world, as you don't really need it outside a hash map. However, sometimes hash map creations gets an overloaded constructor where you can specify your starting capacity in case that you already know your hash map going to be big from the start. The same goes for the load factor.

I'm not sure if we should add the overloaded constructor details, or more methods to the project. There are many that we can add like `addIfAbsent` or `getOrDefault` and so on. But those are just extra quality of life kind of things, while our lesson of importance is to understand the inner workings of a hash map, not necessarily the most evolved one.

>   1. `length` returns the number of stored keys in the hash map.

Reviewer 2: Sounds reasonable. We could probably add the constructor (and other methods) in extra credit section sometime in the future. Good to go for now!

  1. `clear` removes all entries in the hash map.

  1. `keys` returns an array containing all the keys inside the hash map.

  1. `values` returns an array containing all the values.

  1. `entries` returns an array that contains each `key, value` pair. Example: `[[firstKey, firstValue], [secondKey, secondValue]]`

>   Remember that a hash map does not preserve insertion order when you are retrieving your arrays' data, it is normal and expected for keys and values to appear out of the order you inserted them in.

Author: I see where I got mixed up with words, I think I meant to say `when you are retrieving your hash map's data as an array` since this is for the `entries` method which gives us an array.

Let me know if you think just saying `hash map's` data sounds better, or we can word it differently to present the idea

>   Remember that a hash map does not preserve insertion order when you are retrieving your arrays' data, it is normal and expected for keys and values to appear out of the order you inserted them in.

Reviewer 2: I think `hash map's` sound clearer. If you other ideas to present this information, go for it

### Introduction

You already know the magic behind hash maps, now it's time to write your own implementation!

#### Limitation

>   Before we get started, we need to lay down some ground rules. JavaScript's dynamic nature of array allow us to insert and retrieve indexes that is outside our array size range. Example: if we create an array of size `16` to be our buckets size, nothing stopping us from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.

Reviewer: ```suggestion
  Before we get started, we need to lay down some ground rules. JavaScript's dynamic nature of array allow us to insert and retrieve indexes that are outside our array size range. Example: if we create an array of size `16` to be our buckets size, nothing stopping us from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
```

  1. Use the following snippet whenever you access a bucket through an index. We want to throw an error if we try to access an out of bound index:

      ```javascript
      if (index < 0 || index >= buckets.length) {
        throw new Error("Trying to access index out of bound");
      }
      ```

### Assignment

<div class="lesson-content__panel" markdown="1">

  1. Start by creating a `HashMap` class or factory function. It's up to you which you want to use. Then proceed to create the following methods:

  1. `hash` takes a value and produces a hash code with it. We did implement a fairly good `hash` function in the previous lesson. You are free to use that, or if you wish, you can conduct your own research. Beware, this is a deep deep rabbit hole.

      <div class="lesson-note lesson-note--tip" markdown="1">
        Hash maps could accommodate various data types for keys like numbers, strings, objects. But for this project, only handle keys of type strings.
      </div>

>   1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key, if a key already exists then the old value is overwritten.

Reviewer: ```suggestion
  1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key. If a key already exists, then the old value is overwritten.
```

     - Remember to grow your buckets size when it needs to, by calculating if your bucket has reached the `load factor`.

  1. `get` takes one argument as a key and returns the value that is assigned to this key. If key is not found, return `null`.

  1. `has` takes a key as an argument and returns `true` or `false` based on whether or not the key is in the hash map.

  1. `remove` takes a key as argument and removes it from the hash table.

  1. `length` returns the number of stored keys in the hash map.

  1. `clear` removes all entries in the hash map.

  1. `keys` returns an array containing all the keys inside the hash map.

  1. `values` returns an array containing all the values.

  1. `entries` returns an array that contains each `key, value` pair. Example: `[[firstKey, firstValue], [secondKey, secondValue]]`

>   Remember that a hash map does not preserve insertion order when you are retrieving your hash map's data, it is normal and expected for keys and values to appear out of the order you inserted them in.

Reviewer: ```suggestion
  Remember that a hash map does not preserve insertion order when you are retrieving your hash map's data. It is normal and expected for keys and values to appear out of the order you inserted them in.
```
